File: spaceship_crm/scala/spaceship_crm_scala_play_frm/app/repository/UserActivityRepository.scala
/**
 * Author veni.vidi.dev (veni.vidi.dev@gmail.com)
 */
package repository

import java.time.LocalDateTime

import bean.UserActivityBean
import config.DBConfig.db
import frm.entity.UserActivity
import frm.table.ActivityTable.activityQuery
import frm.table.UserActivityTable.userActivityQuery
import frm.table.UserTable.userQuery
import javax.inject.{Inject, Singleton}
import slick.jdbc.PostgresProfile.api._

import scala.concurrent._

@Singleton
class UserActivityRepository @Inject() ()(implicit ec: ExecutionContext) {

  /**
    * query:
    *     select u.id, u.first_name, u.last_name, a.name, a.start_date, a.end_date, ua.created_at
    *     from activity a
    *       left join user_activity ua on a.id = ua.activity_id
    *       left join users u on ua.user_id = u.id
    *
    * @return
    */
  def findAll(): Future[Seq[UserActivityBean]] = {
    val userActivitiesQuery =
      activityQuery
        .join(userActivityQuery).on(_.id === _.activityId)
        .join(userQuery).on(_._2.userId === _.id)

    val action = {
      for {
        userActivityResult <- userActivitiesQuery.result
      } yield {
        userActivityResult.map { row =>
            new UserActivityBean(
              row._2.id,
              row._2.firstName,
              row._2.lastName,
              row._1._1.name,
              row._1._1.startDate,
              row._1._1.endDate,
              row._1._2.createdAt
            )
        }
      }
    }
    db.run(action)
  }

  def insert(activityId: Long, userId: Long): Future[Int] = {
    val insertUserActivity = userActivityQuery += UserActivity(userId, activityId, LocalDateTime.now(), None)
    db.run(insertUserActivity)
  }

  def delete(activityId: Long, userId: Long): Future[Int] = {
    val deleteByIdQuery =
      userActivityQuery
        .filter(ua => ua.userId === userId && ua.activityId === activityId)
          .delete
    db.run(deleteByIdQuery)
  }
}

File: spaceship_crm/java8/spaceship_crm_java8_rest_jooq/src/main/java/com/spaceship/crm/repository/UserActivityRepository.java
/**
 * Author veni.vidi.dev (veni.vidi.dev@gmail.com)
 */
package com.spaceship.crm.repository;

import com.spaceship.crm.bean.UserActivityBean;
import com.spaceship.crm.jooq.Tables;
import com.spaceship.crm.jooq.tables.records.UserActivityRecord;
import org.jooq.DSLContext;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Repository;

import java.sql.Timestamp;
import java.util.List;

import static org.jooq.impl.DSL.row;

//TODO: Framework should be created
@Repository
public class UserActivityRepository {

    @Autowired
    DSLContext dsl;

//    "select u.id, u.first_name, u.last_name, a.name, a.start_date, a.end_date, ua.created_at " +
//            "from user_activity ua " +
//            "   left join users u on ua.user_id = u.id " +
//            "   left join activity a on ua.activity_id = a.id "

    public List<UserActivityBean> findAll() {
        return dsl
                .select(
                    Tables.USERS.ID.as("user_id"),
                    Tables.USERS.FIRST_NAME,
                    Tables.USERS.LAST_NAME,
                    Tables.ACTIVITY.NAME.as("activity_name"),
                    Tables.ACTIVITY.START_DATE.as("start"),
                    Tables.ACTIVITY.END_DATE.as("end"),
                    Tables.USER_ACTIVITY.CREATED_AT.as("added")
                )
                .from(Tables.USER_ACTIVITY)
                    .leftJoin(Tables.USERS).onKey()
                    .leftJoin(Tables.ACTIVITY).onKey()
                .fetch()
                .into(UserActivityBean.class);
    }

    public void insert(Long activityId, Long userId) {
        UserActivityRecord userActivityRecord = dsl.newRecord(Tables.USER_ACTIVITY);
        userActivityRecord.setActivityId(activityId);
        userActivityRecord.setUserId(userId);
        userActivityRecord.setCreatedAt(new Timestamp(System.currentTimeMillis()));
        userActivityRecord.store();
    }

    public void delete(Long activityId, Long userId) {
        dsl
                .deleteFrom(Tables.USER_ACTIVITY)
                .where(
                        row(Tables.USER_ACTIVITY.ACTIVITY_ID, Tables.USER_ACTIVITY.USER_ID)
                                .equal(row(activityId, userId)))
                .execute();
    }
}
